### PR TITLE
Add ignorePureComponents check on 'this' usage

### PR DESCRIPTION
**Pull request by metreniuk**
This PR address the #1901 issue.

### CODE REVIEW

File: tests/lib/rules/prefer-stateless-function.js [hunk 1]
```diff
@@ -345,6 +345,20 @@ ruleTester.run('prefer-stateless-function', rule, {
       errors: [{
         message: 'Component should be written as a pure function'
       }]
+    }, {
+      code: `
+        class Foo extends React.PureComponent {
+          render() {
+            return <div>this.bar</div>;
```
Comment (ljharb) [inline tests/lib/rules/prefer-stateless-function.js]: I'm a bit confused here; how can the linter statically know that there's no `bar = foo` defined in the class or the constructor?

Comment (metreniuk) [reply tests/lib/rules/prefer-stateless-function.js]: I see the problem. I was confused by the docs:
>When true the rule will ignore Components extending from React.PureComponent that use this.props or this.context.

The example that you mentioned leads to the fact that this rule will ignore all `this`'s properties, not only `props` and `context`. Is it accurate?

Comment (ljharb) [reply tests/lib/rules/prefer-stateless-function.js]: If you have a component that doesn't reference props or context, then there's other better mechanisms for anything that hangs off of `this` - so the rule *should* warn on it, and you should refactor to use an SFC.

Comment (alexzherdev) [inline tests/lib/rules/prefer-stateless-function.js]: Is this line missing curlies around `this.bar`?

Comment (ljharb) [reply tests/lib/rules/prefer-stateless-function.js]: good catch, yes.

**Comment by metreniuk** [comment]
I've accidentally removed the fork from which I made the PR.  I will check if I could come with another one.

**Comment by ljharb** [comment]
That's unfortunate.

**Comment by ljharb** [comment]
@metreniuk any chance you'd be able to recreate the PR?

**Comment by metreniuk** [comment]
@ljharb Yes, I will work on it in the begging of the week

**Comment by ljharb** [comment]
Any update?

**Comment by metreniuk** [comment]
Sorry again. I was looking into it and found a positive test case in the master for the issue addressed in this PR
https://github.com/jsx-eslint/eslint-plugin-react/blob/c88c7c497a75988497d14bc492aee9b8cf3b05a3/tests/lib/rules/prefer-stateless-function.js#L158-L167

Is this the intended behaviour or in this case we still want to warn the user to use SFC?

**Comment by ljharb** [comment]
In that case, `this.bar` isn't valid anyways, so I'm not really sure. It's *probably* a bug, and should be `this.props.bar`, in which case we _would_ want to warn and suggest an SFC.

**Comment by metreniuk** [comment]
I'm still confused. The docs says that if `instance property other than this.props and this.context` is found (ex: `this.bar`) then the rule should not warn. I think I misunderstood the rule in the first place. If you have a case that would make sense to cover in this rule, let me know.

**Comment by ljharb** [comment]
yes, that's fair, and I agree with your interpretation.

**Pull request closed without merging**